ts/features/bonus/cgn/saga/networking/bucket/index.ts

```diff
@@ -1,12 +1,38 @@
-import { delay, put } from "redux-saga/effects";
+import { call, put } from "redux-saga/effects";
+import { SagaCallReturnType } from "../../../../../../types/utils";
+import { getNetworkError } from "../../../../../../utils/errors";
+import { readablePrivacyReport } from "../../../../../../utils/reporters";
+import { BackendCgnMerchants } from "../../../api/backendCgnMerchants";
 import { cgnCodeFromBucket } from "../../../store/actions/bucket";
 
 // handle the request for CGN bucket consumption
 export function* cgnBucketConsuption(
-  // TODO Replace with the proper client once defined
-  _: unknown
+  getDiscountBucketCode: ReturnType<
+    typeof BackendCgnMerchants
+  >["getDiscountBucketCode"],
+  cgnCodeFromBucketRequest: ReturnType<typeof cgnCodeFromBucket["request"]>
 ) {
-  // TODO Replace with the try-catch block to properly handle the code consuption API request
-  yield delay(200);
-  yield put(cgnCodeFromBucket.success({ code: "SAMPLECODEFROMBUCKET" }));
+  try {
+    const discountBucketCodeResult: SagaCallReturnType<
+      typeof getDiscountBucketCode
+    > = yield call(getDiscountBucketCode, {
+      discountId: cgnCodeFromBucketRequest.payload
+    });
+    if (discountBucketCodeResult.isRight()) {
+      if (discountBucketCodeResult.value.status === 200) {
+        yield put(
+          cgnCodeFromBucket.success(discountBucketCodeResult.value.value)
+        );
+        return;
+      } else {
+        throw new Error(
```

Comment:
I don't think throwing an error to manage it in the catch is a best practice.
What do you think about putting a failure action instead?

Reply:
This was the original code for this saga, as you can see putting multiple times the failure action caused a repetition of code, in this way i centralized the dispatch of failure action, I think they're both good ways but I don't have a strong position about which would be better to use.

```typescript
  try {
    const discountBucketCodeResult: SagaCallReturnType<
      typeof getDiscountBucketCode
    > = yield call(getDiscountBucketCode, {
      discountId: cgnCodeFromBucketRequest.payload
    });
    if (discountBucketCodeResult.isRight()) {
      if (discountBucketCodeResult.value.status === 200) {
        yield put(
          cgnCodeFromBucket.success(discountBucketCodeResult.value.value)
        );
      } else {
        yield put(
          cgnCodeFromBucket.failure(
            getNetworkError(
              new Error(
                `response status ${discountBucketCodeResult.value.status}`
              )
            )
          )
        );
      }
    } else {
      yield put(
        cgnCodeFromBucket.failure(
          getNetworkError(
            new Error(readablePrivacyReport(discountBucketCodeResult.value))
          )
        )
      );
    }
  } catch (e) {
    yield put(cgnCodeFromBucket.failure(getNetworkError(e)));
  }
```

Reply:
I am not a fundamentalist but semantically I don't know if throwing an error it's as readable as dispatching the action.

> Maybe since we should properly handle error cases through https://pagopa.atlassian.net/browse/IOACGN-21 we can go through errors in an appropriate way what do you think?

Yes, works for me